torch/distributions.py

```diff
@@ -178,16 +189,59 @@ def sample(self):
         return torch.normal(self.mean, self.std)
 
     def sample_n(self, n):
-        # cleanly expand float or Tensor or Variable parameters
-        def expand(v):
-            if isinstance(v, Number):
-                return torch.Tensor([v]).expand(n, 1)
-            else:
-                return v.expand(n, *v.size())
-        return torch.normal(expand(self.mean), expand(self.std))
+        return torch.normal(_expand_n(self.mean, n), _expand_n(self.std, n))
 
     def log_prob(self, value):
         # compute the variance
         var = (self.std ** 2)
         log_std = math.log(self.std) if isinstance(self.std, Number) else self.std.log()
         return -((value - self.mean) ** 2) / (2 * var) - log_std - math.log(math.sqrt(2 * math.pi))
+
+
+def _standard_gamma(alpha):
+    if not isinstance(alpha, Variable):
+        return torch._C._standard_gamma(alpha)
+    return Variable(torch._C._standard_gamma(alpha.data))
+
+
+class Gamma(Distribution):
+    r"""
+    Creates a Gamma distribution parameterized by shape `alpha` and rate `beta`.
+
+    Example::
+
+        >>> m = Gamma(torch.Tensor([1.0]), torch.Tensor([1.0]))
+        >>> m.sample()  # Gamma distributed with shape alpha=1 and rate beta=1
+         0.1046
+        [torch.FloatTensor of size 1]
+
+    Args:
+        alpha (float or Tensor or Variable): shape parameter of the distribution
+        beta (float or Tensor or Variable): rate = 1 / scale of the distribution
+    """
+
+    def __init__(self, alpha, beta):
+        # TODO handle (Variable, Number) cases
+        alpha_num = isinstance(alpha, Number)
+        beta_num = isinstance(beta, Number)
+        if alpha_num and not beta_num:
+            alpha = beta.new(beta.size()).fill_(alpha)
+        elif not alpha_num and beta_num:
+            beta = alpha.new(alpha.size()).fill_(beta)
+        elif alpha_num and beta_num:
+            alpha, beta = torch.Tensor([alpha]), torch.Tensor([beta])
+        else:
+            alpha = alpha.expand_as(beta)
```

Comment:
If you want to support broadcasting, I think you have to do it both ways here (i.e. broadcast the smaller thing to the larger one). Consider this case: `beta.size() == [2]`, `alpha.size() == [4, 4, 2]`.

Reply:
One-sided broadcasting actually works fine in this case. We only need to expand `alpha` because `alpha.size()` determines the number of `_standard_gamma()` samples that are drawn. `beta` can remain small, and will be broadcast correctly during scaling: `_standard_gamma(alpha) / beta`. (The purpose of commit 7387792 is to ensure that enough samples are drawn, since samply-copying is not the correct behavior.)

Reply:
The problem is that this line will raise an error if someone passes in a small beta and large alpha, so you won't even get to sampling.